sleap/nn/inference.py

```diff
@@ -1695,49 +1895,42 @@ def make_labeled_frames_from_generator(self, generator, data_provider):
 
         return predicted_frames
 
-    def predict_generator(self, data_provider: Provider):
-
-        if self.pipeline is None:
-            self.make_pipeline()
-        if self.inference_model is None:
-            self._initialize_inference_model()
-
-        self.pipeline.providers = [data_provider]
-
-        for ex in self.pipeline.make_dataset():
-            preds = self.inference_model.predict(ex)
-
-            ex["instance_peaks"] = [
-                x[:n] for x, n in zip(preds["instance_peaks"], preds["n_valid"])
-            ]
-            ex["instance_peak_vals"] = [
-                x[:n] for x, n in zip(preds["instance_peak_vals"], preds["n_valid"])
-            ]
-            ex["instance_scores"] = [
-                x[:n] for x, n in zip(preds["instance_scores"], preds["n_valid"])
-            ]
-            yield ex
-
     def predict(
         self,
-        data_provider: Provider,
-        make_instances: bool = True,
-        make_labels: bool = False,
-    ):
-        if isinstance(data_provider, sleap.Labels):
-            data_provider = LabelsReader(data_provider)
-        elif isinstance(data_provider, sleap.Video):
-            data_provider = VideoReader(data_provider)
-        generator = self.predict_generator(data_provider)
+        data: Union[Provider, sleap.Labels, sleap.Video],
+        make_labels: bool = True,
+    ) -> Union[List[Dict[str, np.ndarray]], sleap.Labels]:
+        """Run inference and tracking on a data source.
 
-        if make_instances or make_labels:
-            lfs = self.make_labeled_frames_from_generator(generator, data_provider)
-            if make_labels:
-                return sleap.Labels(lfs)
-            else:
-                return lfs
+        Args:
+            data: A `sleap.pipelines.Provider`, `sleap.Labels` or `sleap.Video` to
+                run inference over.
+            make_labels: If `True` (the default), returns a `sleap.Labels` instance with
+                `sleap.PredictedInstance`s. If `False`, just return a list of
+                dictionaries containing the raw arrays returned by the inference model.
 
-        return list(generator)
+        Returns:
+            A `sleap.Labels` with `sleap.PredictedInstance`s if `make_labels` is `True`,
+            otherwise a list of dictionaries containing batches of numpy arrays with the
+            raw results.
+        """
+        # Create provider if necessary.
+        if isinstance(data, sleap.Labels):
```

Comment:
I know this is not new, but just curious how is this casting helpful here?

Reply:
This is a convenience shortcut. Say you have some `sleap.Labels` you want to predict on. This saves an unnecessary step:
```py
predictor.predict(labels)
```
as compared to:
```py
predictor.predict(sleap.pipelines.LabelsReader(labels))
```